### ISSUE TITLE
Getting sbt test to work

### ISSUE THREAD

**Issue by mhyee** [flix/flix#168]
I ran the ScalaTests on the command line without SBT. By default, it runs the tests sequentially. I'm still getting the failure, so it's probably something to do with the class loader. (The fact that the code generator uses reflection and has its own class loader is probably why this is so messy.)

Here's the error:

```
*** RUN ABORTED ***
  java.lang.NoClassDefFoundError: Lca/uwaterloo/flix/api/Flix;
  at java.lang.Class.getDeclaredFields0(Native Method)
  at java.lang.Class.privateGetDeclaredFields(Class.java:2583)
  at java.lang.Class.getField0(Class.java:2975)
  at java.lang.Class.getField(Class.java:1701)
  at ca.uwaterloo.flix.language.phase.LoadBytecode$$anonfun$5.apply(LoadBytecode.scala:99)
  at ca.uwaterloo.flix.language.phase.LoadBytecode$$anonfun$5.apply(LoadBytecode.scala:92)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
  at scala.collection.immutable.Map$Map1.foreach(Map.scala:116)
  at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
  ...
  Cause: java.lang.ClassNotFoundException: ca.uwaterloo.flix.api.Flix
  at java.lang.ClassLoader.findClass(ClassLoader.java:530)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
  at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
  at java.lang.Class.getDeclaredFields0(Native Method)
  at java.lang.Class.privateGetDeclaredFields(Class.java:2583)
  at java.lang.Class.getField0(Class.java:2975)
  at java.lang.Class.getField(Class.java:1701)
  at ca.uwaterloo.flix.language.phase.LoadBytecode$$anonfun$5.apply(LoadBytecode.scala:99)
  at ca.uwaterloo.flix.language.phase.LoadBytecode$$anonfun$5.apply(LoadBytecode.scala:92)
  at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
  ...
```

Running ScalaTest from the command-line is a pain. If I use IntelliJ to package everything into a single jar, I can do:

`scala -cp lib/org.scalatest-2.2.6.jar:out/flix.jar org.scalatest.tools.Runner -R "target/scala-2.11/test-classes/" -o -s ca.uwaterloo.flix.TestAll`

Otherwise I have to put all the jars on the classpath:

`scala -cp lib/org.json4s-ast-3.2.9.jar:lib/com.github.shapeless-2.3.0.jar:lib/com.microsoft.z3.jar:lib/org.json4s-ast-3.2.9.jar:lib/org.json4s-core-3.2.11.jar:lib/org.json4s-native-3.2.11.jar:lib/org.objectweb.asm-all-5.0.4.jar:lib/org.parboiled-2.1.2.jar:lib/org.scalatest-2.2.6.jar org.scalatest.tools.Runner -R "target/scala-2.11/test-classes/ target/scala-2.11/classes/" -o -s ca.uwaterloo.flix.TestAll`

**Comment by magnus-madsen** [comment]
I can confirm this also happens with Ant, if you don't fork the VM.

**Comment by paulp** [comment]
Here's a way to run the tests from the command line which doesn't require manually dredging paths.
```bash
java -cp $( sbt 'export test:fullClasspath' | egrep -v '^\S*info' ) \
  org.scalatest.tools.Runner \
  -R target/scala-2.11/test-classes \
  -o -s ca.uwaterloo.flix.TestAll
```
You should be able to run the tests using scala if you pass `-nobootcp` to prevent it from putting the scala jars on the boot classpath, except that it seems the option stopped working years ago. I reported it as [SI-9833](https://issues.scala-lang.org/browse/SI-9833).

**Comment by paulp** [comment]
And having already produced the cart, here's the horse: most of the tests pass fine if you add `fork in Test := true` in build.sbt.

**Comment by paulp** [comment]
Ok, on OSX all tests pass after adding `fork in Test := true` and invoking sbt as
`DYLD_LIBRARY_PATH=/usr/local/opt/z3/lib sbt test`. The environment variable works where `-Djava.library.path` does not, presumably because the environment variable survives across the fork whereas the command line option does not. Maybe there's a simple way in sbt to address that; the annoying way will involve using sbt's fork api.

**Comment by paulp** [comment]
> (3) Will you look into how we can make sbt test only run TestAll?

@magnus-madsen This appears to work:
```
test :=  (testOnly in Test).toTask(" *.TestAll").value
```

**Comment by paulp** [comment]
Okay, these changes are submitted as #180.

**Comment by magnus-madsen** [comment]
Oh wow! Hey Paul. Thanks for your interest in Flix. And thanks for the fix!

How did you find out about us?

Anyway, I would be happy to merge your changes. Can you either add yourself to CONTRIBUTORS.md or state here that you are making your changes available under Apache 2.0?

**Comment by paulp** [comment]
I added myself, despite the little shiver of self-promotion it sent down my spine. It's inspired to tie the license acknowledgement to the contributors file. I think I'll steal that idea.

@sirinath pointed me at you on twitter.
